### PR TITLE
compcert 3.0

### PR DESCRIPTION
**Pull request by beurdouche**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by beurdouche** [comment]
? @ilovezfs

**Comment by ilovezfs** [comment]
@beurdouche sorry for not commenting right away. So ... a couple things. First, we never put `revision 0`. The counting of revisions does start from 0, but we only add an explicit revision when it's >= 1.

Second, are you sure we can flat out replace the ia32-macosx with x86_64-macosx, and don't need to build both or make one optional and the other default? I saw in the Changelog that it said:

>The x86 port of CompCert was extended to produce x86-64 bit code in
  addition to the original x86-32 bit (IA32) code.  (This is the first
  instantiation of the new support for 64-bit targets described
  above.) Support for x86-64 is currently available for Linux and MacOS X.
  (Run the configure script with 'x86_64-linux' or 'x86_64-macosx'.)
  This is an early port: several ABI incompatibilities remain.

(http://compcert.inria.fr/release/Changelog)

**Comment by beurdouche** [comment]
@ilovezfs no worries. 
After thinking a bit more about it, I think you are right. It might be more reasonable to keep `ia32-macosx` as the default for this release and offer the option for building `x86_64`.
I'll discuss with Xavier Leroy (the author) to make sure that we can switch default and option in a future release.
Also I just noticed a bug in the version number (hasn't been incremented to 3.0.0) and reported it upstream. Maybe we can do a patch just to fix the VERSION file ?
WDYT ?

**Comment by ilovezfs** [comment]
Yup. All sounds correct to me. If the version number does get patched upstream right away, but a new tarball isn't published, then it would be best to apply the commit as a patch, but if not an `inreplace` will work.

**Comment by beurdouche** [comment]
Ok, the version problem has been fixed and the release has been re-tagged upstream, so no patch necessary.

Remains: 
[] updating the script with the new tarball/hash
[] creating the option to build with ia32 or x86_64

I'll try to do that tomorrow except if someone manages to do it before..

**Comment by ilovezfs** [comment]
@beurdouche did you get a chance to talk with Xavier?

**Comment by beurdouche** [comment]
@ilovezfs Ok I discussed with Xavier by email. He suggested to separate Compcert32 (more stable) and Compcert64 (more experimental) for now, so I proceeded by adding an option to compile `x86_64` if the user chooses it. Please note that while it seems to work properly I am not super familiar with options for Homebrew and, in particular, naming conventions. The PR is currently ready to be merged from my perspective, please don't hesitate to make adjustments where you see fit ;)

**Comment by beurdouche** [comment]
Ps: just squashed the commits

File: Formula/compcert.rb [hunk 1]
```diff
@@ -36,8 +31,15 @@ def install
     (Pathname.pwd/"opamroot").mkpath
     system "opam", "init", "--no-setup"
     system "opam", "install", "coq=8.5.3"
-    system "opam", "config", "exec", "--",
-           "./configure", "-prefix", prefix, "ia32-macosx"
+
+    if build.without? "config-x86_64"
```
Comment (MikeMcQuaid) [inline Formula/compcert.rb]: Use `build.with?` instead; it's a little easier to follow.

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

**Comment by MikeMcQuaid** [comment]
> Compcert64 (more experimental) for now, so I proceeded by adding an option to compile x86_64 if the user chooses it.

Maybe it's worth removing the option unless you personally use it?

**Comment by ilovezfs** [comment]
The option is the cool new toy.

**Comment by beurdouche** [comment]
@MikeMcQuaid @ilovezfs, I actually suspect that the option will be very used, yes :)
JustChanged the code to use `build.with ?` instead, thank you for the suggestion.

**Comment by MikeMcQuaid** [comment]
> @MikeMcQuaid @ilovezfs, I actually suspect that the option will be very used, yes :)

@beurdouche Will you personally be using it or not?

**Comment by beurdouche** [comment]
@MikeMcQuaid Yes, I am using it, as well as half of the lab.

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by MikeMcQuaid** [comment]
@beurdouche In that case I might be in favour for saying it should just be the default?

**Comment by ilovezfs** [comment]
@MikeMcQuaid upstream has basically explicitly said that it's not considered stable yet and there are API incompatibilities.

**Comment by beurdouche** [comment]
@MikeMcQuaid we discussed it but we think it is too early. We will probably change the default in 3.1 or 3.2 after another discussion with the authors.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by beurdouche** [comment]
@BrewTestBot test this please

**Comment by beurdouche** [comment]
@ilovezfs Good to go ;)

**Comment by MikeMcQuaid** [comment]
Thanks again for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Comment by beurdouche** [comment]
@ilovezfs @MikeMcQuaid Thanks a lot to both of you ! ;)

**Pull request closed without merging**